Sort resources by name

abstractHunter opened the pull request:
Hello, I added a python script that will automatically sort resource files links. After inserting new material links, you can just run the script and the file will be sorted.
I'm open to any remarks and changes if needed.

Commit:
Add script to sort resource and learning material links alphabetically

Commit:
Sort printf/resources.md and simple-shell/learning-material.md links

geoffreylgv reviewed (commented):
Hi @milkiyd, did you use python code to sort automatically all links in the files?
(I mean I'm seeing it, but just to confirm)

geoffreylgv reviewed (changes requested):
Here we have the learning resource starting by `S` but, still stay on the bottom.

geoffreylgv commented on sort_resource_links.py:
```diff
@@ -0,0 +1,149 @@
+"""
+    If you added a new resource to an existing resources.md or learning-material.md file, just run this script directly.
+    If you added a new resources.md or learning-material.md file, add the path to the file to the RESOURCE_FILE_PATHS list
+    below and then run this script.
+
+    To run this script, open a terminal or command prompt and navigate to the directory that contains this script.
+
+    If you're using a Mac or a Linux distribution, you can run this script from the terminal by typing the following command:
+        python3 sort_resource_links.py
+    If you're using Windows, you can run this script from the command prompt by typing the following command:
+        python sort_resource_links.py
+
+    Note 1: You must have Python 3 installed on your computer to run this script. If you don't have Python 3 installed,
+    you can download it from https://www.python.org/downloads/. Make sure you download the version for your operating
+    system (Windows, Mac, or Linux).
+
+    Note 2: This script assumes that it is at the root of the repository. If you move this script to a different location,
+    you will need to update the RESOURCE_FILE_PATHS list below to include the correct paths to the resources.md and
+    learning-material.md files.
```
Hello @milkiyd, we should write a 'guide' to explain this instead of let it here in the `.py` code

geoffreylgv commented on sort_resource_links.py:
Let's keep the name of the program simple, what do you think about `sort-resources.py`
The instruction of write by alphabetical should be added to the contributing guidelines tho, I'll manage that soon

abstractHunter replied:
Yes, i also prefer this name.
I also agree to have the instructions in a separate file (I didn't want to modify or add many files before you agree on what I were doing).

geoffreylgv reviewed (commented):
Hi @milkiyd Sorry for the delay, I was and still working on a project `(simple shell project)`,
You are doing a great work and that is nice work you did, let's make it nicer 💯

abstractHunter commented:
> Hi @milkiyd, did you use python code to sort automatically all links in the files? (I mean I'm seeing it, but just to confirm)

Yes it is python



> Here we have the learning resource starting by `S` but, still stay on the bottom.

It may due to the fact that the resource name is starting with a lower case 's' which have an ascii code (115) greater than the uppwercase 'W' (87). Sorry i didn't checked that before commiting.

One fix could be to make sure that all resource names start with an uppercase letter, or I could try to handle these cases directly in the code.

geoffreylgv commented:
> > Hi @milkiyd, did you use python code to sort automatically all links in the files? (I mean I'm seeing it, but just to confirm)
> 
> Yes it is python
> 
Alright, nice

> > Here we have the learning resource starting by `S` but, still stay on the bottom.
> 
> It may due to the fact that the resource name is starting with a lower case 's' which have an ascii code (115) greater than the uppwercase 'W' (87). Sorry i didn't checked that before commiting.
> 
> One fix could be to make sure that all resource names start with an uppercase letter, or I could try to handle these cases directly in the code.

No worries @milkiyd, always start with capital letter any title or sentence, but for enhancement of the code, I agree that if possible, because it'll serve you in future project I guess and sure, so you can handle if it's a lower letter or not.

Commit:
Update resources sorting script to be case insensitive | Sort simple-shell/learning-material.md

Commit:
Update resources sorting script to handle resources containing h4 headings | Update CONTRIBUTING.md file

abstractHunter commented:
Hello @geoffreylgv , I updated the script according to your reviews, and added the feature to handle resources containing subsections. 
The CONTRIBUTING.md file has also been updated.

geoffreylgv commented:
> Hello @geoffreylgv , I updated the script according to your reviews, and added the feature to handle resources containing subsections. The CONTRIBUTING.md file has also been updated.

Hi @milkiyd, I will look into it soon, well done

geoffreylgv commented on CONTRIBUTING.md:
```diff
@@ -125,3 +125,27 @@ Soon I'll be merging all your changes into the main branch of this project. You
 Congrats! You just completed the standard _fork -> clone -> edit -> pull request_ workflow that you'll often encounter as a contributor!
 
 Celebrate your contribution and share it with your friends and followers on Twitter and tag the team.
+
+### Adding resources
+- If you added a new resource to an existing `resources.md` or `learning-material.md` file, just run `sort-resources.py` directly.
+    
+- If you added a new `resources.md` or `learning-material.md` file, add the path to the file to the RESOURCE_FILE_PATHS list below and then run `sort-resources.py`.
+
+To run this script, open a terminal or command prompt and navigate to the directory that contains this script.
+
+If you're using a Mac or a Linux distribution, you can run this script from the terminal by typing the following command: `python3 sort_resource_links.py`
+
+If you're using Windows, you can run this script from the command prompt by typing the following command:
+`python sort_resource_links.py`
+
+Note 1: The links must be in the following format: `- [Link text](link URL)`. If you use a different format, the script will not work. 
+
+Note 2: The heading for the sections must use a h4 heading (`####`). If you use a different heading, the script will not work.
+
+Note 3: You must have Python 3 installed on your computer to run this script. If you don't have 
+Python 3 installed, you can download it from https://www.python.org/downloads/. Make sure you 
+download the version for your operating system (Windows, Mac, or Linux).
+
+Note 4: This script assumes that it is at the root of the repository. If you move this script to 
+a different location, you will need to update the RESOURCE_FILE_PATHS list below to include the 
+correct paths to the resources.md and learning-material.md files.
```
Awesome @milkiyd, even if we will update it, it's much better

geoffreylgv commented on dsa/resources/resources.md:
```diff
@@ -1,24 +1,24 @@
-- [Project Euler](https://projecteuler.net/)
-- [Coding game](https://www.codingame.com/)
-- [CodeWars](https://www.codewars.com/)
+#### Courses articles blogs
+
+- [Collection of Data Structure and Algorithms learning material - Google](https://techdevguide.withgoogle.com/paths/data-structures-and-algorithms/)
+- [Data Structure - Geeks for Geeks](https://www.geeksforgeeks.org/data-structures/)
+- [Data Structure and Algorithms - CS50 Havard University](https://cs50.harvard.edu/law/2019/weeks/3/)
+- [Data Structure and Algorithms - Freecodecamp](https://www.freecodecamp.org/news/learn-data-structures-and-algorithms/)
+- [Data Structure and Algorithms - Tutorialspoint](https://www.tutorialspoint.com/data_structures_algorithms/index.htm)
+- [Data Structure Visualizations - University of San Francisco](https://www.cs.usfca.edu/~galles/visualization/Algorithms.html)
+
+#### Videos tutorials
+
+- [Algorithms and Data Structures Tutorial - Full Course for Beginners - Freecodecamp](https://www.youtube.com/watch?v=8hly31xKli0)
+- [Complete Data Structures and Algorithms - Language: C - **Jenny's Lectures CS IT**](https://www.youtube.com/watch?v=AT14lCXuMKI&list=PLdo5W4Nhv31bbKJzrsKfMpo_grxuLl8LU)
+- [Data Structures - Full Course Using C and C++](https://www.youtube.com/watch?v=B31LgI4Y4DQ&t=8488s)
+- [Programming & Data Structures - Language C - **Neso Academy**](https://www.youtube.com/watch?v=xLetJpcjHS0&list=PLBlnK6fEyqRj9lld8sWIUNwlKfdUoPd1Y)
+
+#### Where to have the assessment ?
+
+- [CoderByte](https://coderbyte.com/)
+- [CodeWars](https://www.codewars.com/)
+- [Coding game](https://www.codingame.com/)
+- [HankerRank](https://www.hackerrank.com/dashboard)
+- [LeetCode](https://leetcode.com/)
+- [Project Euler](https://projecteuler.net/)
```
Yeah, nice work here !

geoffreylgv commented on sort-resources.py:
Since is nice now, I'll politely merge it, thank you, @milkiyd, for the great work

geoffreylgv reviewed (approved):
Congrats Daniel, you are the big one.

`"The struggle you’re in today is developing the strengths you need tomorrow. Don’t give up" `
